### PR TITLE
Add i2c dump filtering

### PR DESCRIPTION
**Pull request by barnas-michal**
This PR changes format of the i2c messages dump, to align the messages for easier comparison, and in case of device emulators, it's using the target device name instead of generic "emul".
It also adds a possibility to filter which devices transactions should be dumped. If someone is debugging the i2c communication, the log being flooded with other devices' transactions make it hard to debug. The logged devices
should be specified in the device-tree using the compatible string.

Example of device-tree node:
    i2c-dump-filter {
            compatible = "zephyr,i2c-dump-filter";
            devices = < &display0 >, < &sensor3 >;
    };

### CODE REVIEW

**Review by galak — DISMISSED** [review]
I'm not fan of this as a devicetree node.  Maybe a property under a controller or something.

**Comment by semihalf-barnas-michal** [comment]
> I'm not fan of this as a devicetree node. Maybe a property under a controller or something.

I am not sure if I would be able to easily iterate over all devices with some property set. It would also require to add the optional property to all possible i2c devices.
I know that this node doesn't represent any physical device, but I think it is the easiest and most readable approach to this functionality.

**Comment by tmon-nordic** [comment]
> I know that this node doesn't represent any physical device, but I think it is the easiest and most readable approach to this functionality.

I agree that this is the easiest approach. However, there is constantly a disagreement (not only in Zephyr, but also in Linux kernel) what constitutes a devicetree and what does not. Linux kernel tends to be stricter about not using devicetree for configuration purposes.

Unfortunately I cannot find my older comments about the same issue, but this is perfect example of a place where some "devicetree-like-configuration" would make sense. Maybe we should group such use cases into some issue?

**Review by teburd — DISMISSED** [review]
I like the way this works.

I was under the impression that zephyr compatibles were understood to be zephyr/virtual-ish like devices and configuration.

I don't see how this would easily be done in Kconfig. It wants structured data as configuration and DT is the tool Zephyr has for that.

**Review by aaronemassey — DISMISSED** [review]
@semihalf-barnas-michal 
> I know that this node doesn't represent any physical device, but I think it is the easiest and most readable approach to this functionality.

@tmon-nordic 
> I agree that this is the easiest approach. 

In my opinion, this is really similar to the first acknowledged type of **exception** scenario listed in Zephyr's docs [comparing DeviceTree with KConfig](https://docs.zephyrproject.org/latest/build/dts/dt-vs-kconfig.html). So it's acceptable to me as devicetree mechanism.

**Review by keith-zephyr — DISMISSED** [review]
Non-blocking nits from me.

**Comment by keith-zephyr** [comment]
> > I'm not fan of this as a devicetree node. Maybe a property under a controller or something.
> 
> I am not sure if I would be able to easily iterate over all devices with some property set. It would also require to add the optional property to all possible i2c devices. I know that this node doesn't represent any physical device, but I think it is the easiest and most readable approach to this functionality.

The GPIO hogs driver provides an example for iterating over all devicetree nodes that have the `gpio-controller` property. Something similar could be done here by adding properties to the i2c-controller.yaml base class.  But considering this is a debug feature I like the single devicetree node that collects up the list of I2C devices to log.

File: drivers/i2c/Kconfig
Comment (semihalf-barnas-michal) [reply drivers/i2c/Kconfig]: Done

File: drivers/i2c/Kconfig
Comment (semihalf-barnas-michal) [reply drivers/i2c/Kconfig]: Done

**Comment by semihalf-barnas-michal** [comment]
> hogs driver provides an example for iterating over all devicetree nodes

I've taken a look at it. The difference is that gpio controllers include this yaml, however the i2c filter should be having a reference to device on the i2c bus, not the controller. It probably may have a child-binding, but then there could be a problem with devices that are specified using multiple children due to usage of more than one address on the bus.

**Review by keith-zephyr — DISMISSED** [review]

**Review by teburd — DISMISSED** [review]

**Review by aaronemassey — DISMISSED** [review]

**Comment by semihalf-barnas-michal** [comment]
Is there anything more needed to do before merge?

**Comment by teburd** [comment]
@galak If you could take another look please that would be helpful here

**Comment by teburd** [comment]
@semihalf-barnas-michal Perhaps it would be beneficial to document this debug feature in the i2c bus docs somewhere, as an opt in feature to dump messages, or dump messages with filtering abilities using a dts node.

At the moment someone would need to read the source code to understand how to use this feature.

**Comment by semihalf-barnas-michal** [comment]
> in the i2c bus docs somewhere

I've extracted some of the debugging documentation to another file and added the documentation about this feature there. If you do not like it, I can revert and just add the documentation to the i2c doc file.

**Comment by teburd** [comment]
> > in the i2c bus docs somewhere
> 
> I've extracted some of the debugging documentation to another file and added the documentation about this feature there. If you do not like it, I can revert and just add the documentation to the i2c doc file. I see that there is another file, services/debug which may be in a conflict to my extraction... However I am not sure if the I2C documentation is suitable for this, since it is very long and mostly describing the I2C interface and API...

Great thank you, I think as long as there's some documentation somewhere on it, it will be helpful. If it turns out to be in the wrong place that's easily fixed.

File: doc/develop/debug/index.rst
Comment (keith-zephyr) [inline doc/develop/debug/index.rst:292]:
```suggestion
There is a possibility to log all or some of the I2C transactions done by the application.
```

File: doc/develop/debug/index.rst
Comment (keith-zephyr) [inline doc/develop/debug/index.rst:295]:
```suggestion
uses the ``LOG_DBG`` function to print the contents so the  :kconfig:option:`CONFIG_I2C_LOG_LEVEL_DBG` option must also be enabled.
```

File: doc/develop/debug/index.rst
Comment (keith-zephyr) [inline doc/develop/debug/index.rst:306]:
```suggestion
The first line indicates the I2C controller and the target address of the transaction.
```

File: doc/develop/debug/index.rst
Comment (keith-zephyr) [inline doc/develop/debug/index.rst:307]:
```suggestion
In above example, the I2C controller is named ``io_i2c_ctrl7_port0`` and the target device address
```

File: doc/develop/debug/index.rst
Comment (keith-zephyr) [inline doc/develop/debug/index.rst:315]:
```suggestion
Next lines contain messages, both sent and received.  The contents of write messages is always shown, while the content of read messages is controlled by a parameter to the function ``i2c_dump_msgs_rw``. This function is available for use

```

File: doc/develop/debug/index.rst
Comment (keith-zephyr) [inline doc/develop/debug/index.rst:316]:
```suggestion
by user, but is also called internally by ``i2c_transfer`` API function with read content dump enabled.
```

File: doc/develop/debug/index.rst
Comment (keith-zephyr) [inline doc/develop/debug/index.rst:340]:
```suggestion
By default, all I2C communication is logged between all I2C controllers and I2C targets.
It may litter the log with unrelated devices and make it difficult to effectively debug the
communication with a device of interest.

Enable the Kconfig option :kconfig:option:`CONFIG_I2C_DUMP_MESSAGES_FILTER` to create an allowlist of I2C targets to log. The allowlist of devices is configured using the devicetree, for example::
```

File: doc/develop/debug/index.rst
Comment (keith-zephyr) [inline doc/develop/debug/index.rst:353]: After reviewing the documentation, would it be clearer to call this option `CONFIG_I2C_DUMP_MESSAGES_ALLOWLIST` and the compatible `zephyr,i2c-dump-allowlist`?

Comment (semihalf-barnas-michal) [reply doc/develop/debug/index.rst:353]: Yes, it will be a better name for this feature. I will change that and update this PR.

**Review by keith-zephyr — APPROVED** [review]

**Review by teburd — APPROVED** [review]

File: doc/develop/application/index.rst
Comment (teburd) [inline doc/develop/application/index.rst:434]: This seems like a stray change? Approved already but noticed after

File: doc/develop/modules.rst
Comment (teburd) [inline doc/develop/modules.rst:279]: why is this being single quoted

**Review by teburd — CHANGES_REQUESTED** [review]
Some of the single quoted additions to the docs don't make much sense to me at a quick glance

File: doc/develop/application/index.rst
Comment (kartben) [reply doc/develop/application/index.rst:434]: it's an external link

https://github.com/zephyrproject-rtos/zephyr/blob/d670f1ba6ecd7988c4220025513f55895c40f6c4/doc/develop/application/index.rst?plain=1#L1291

https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#external-links

**Review by teburd — APPROVED** [review]

**Pull request merged**
